rclcpp/multithread_executor/multithread_executor.cpp

```diff
@@ -0,0 +1,188 @@
+     * They don't really do much on their own, but they have to exist in order to
+     * assign callbacks to them. They're also what the executor looks for when trying to run multiple threads
+     */
+    callback_group_subscriber1_ = this->create_callback_group(
+      rclcpp::callback_group::CallbackGroupType::MutuallyExclusive);
+    callback_group_subscriber2_ = this->create_callback_group(
+      rclcpp::callback_group::CallbackGroupType::MutuallyExclusive);
+
+    // Each of these callback groups is basically a thread
+    // Everything assigned to one of them gets bundled into the same thread
+    auto sub1_opt = rclcpp::SubscriptionOptions();
+    sub1_opt.callback_group = callback_group_subscriber1_;
+    auto sub2_opt = rclcpp::SubscriptionOptions();
+    sub2_opt.callback_group = callback_group_subscriber2_;
+
+    subscription1_ = this->create_subscription<std_msgs::msg::String>(
+      "topic",
+      rclcpp::QoS(10),
+      // std::bind is sort of C++'s way of passing a function
+      // If you're used to function-passing, skip these comments
+      std::bind(
+        &DualThreadedNode::subscriber1_cb,  // First parameter is a reference to the function
+        this,                               // What the function should be bound to
+        std::placeholders::_1),             // At this point we're not positive of all the
+                                            // parameters being passed
+                                            // So we just put a generic placeholder
+                                            // into the binder
+                                            // (since we know we need ONE parameter)
+      sub1_opt);                  // This is where we set the callback group.
+                                  // This subscription will run with callback group subscriber1
+
+    subscription2_ = this->create_subscription<std_msgs::msg::String>(
+      "topic",
+      rclcpp::QoS(10),
+      std::bind(
+        &DualThreadedNode::subscriber2_cb,
+        this,
+        std::placeholders::_1),
+      sub2_opt);
+  }
+
+private:
+  /**
+   * Simple function for generating a timestamp
+   * Used for somewhat ineffectually demonstrating that the multithreading doesn't cripple performace
+   */
+  std::string timing_string()
+  {
+    rclcpp::Time time = this->now();
+    return std::to_string(time.nanoseconds());
+  }
+
+  /**
+   * A small convenience function for converting a thread ID to a string
+   */
+  std::string string_thread_id(std::thread::id id)
+  {
+    auto hashed = std::hash<std::thread::id>()(id);
+    return std::to_string(hashed);
+  }
+
+  /**
+   * Every time the Publisher publishes something, all subscribers to the topic get poked
+   * This function gets called when Subscriber1 is poked (due to the std::bind we used when defining it)
+   */
+  void subscriber1_cb(const std_msgs::msg::String::SharedPtr msg)
+  {
+    auto message_recieved_at = timing_string();
```

Comment:
Nit: `message_recieved_at` -> `message_received_at`

(there are a couple of instances of this below as well)